logback-core/src/test/java/ch/qos/logback/core/net/server/ConcurrentServerRunnerTest.java

```diff
@@ -73,73 +94,74 @@ public void testStartStop() throws Exception {
-      while (retries-- > 0 && !client.isRunning()) {
-        client.wait(SHORT_DELAY);
-      }
-    }
+  public void testLifeCycleOfOneClient() throws Exception {
+    listener.addSocket(new Socket());
     assertTrue(runner.awaitRunState(true, DELAY));
+    waitForRunnerClientListSizeToEqual(1);
+    WaitingClient client = runner.clientList.get(0);
+    waitForClientToStart(client);
     client.close();
+    waitForClientToStop(client);
     runner.stop();
   }
 
+
   @Test
   public void testRunManyClients() throws Exception {
-    executor.execute(runner);
-    int count = 10;
-    while (count-- > 0) {
-      MockClient client = new MockClient();
-      listener.addClient(client);
-      int retries = DELAY / SHORT_DELAY;
-      synchronized (client) {
-        while (retries-- > 0 && !client.isRunning()) {
-          client.wait(SHORT_DELAY);
-        }
-      }
-      assertTrue(runner.awaitRunState(true, DELAY));
+    int clientCount = 10;
+
+    for (int i = 0; i < clientCount; i++) {
+      listener.addSocket(new Socket());
     }
+
+    waitForRunnerClientListSizeToEqual(clientCount);
+    for (WaitingClient client : runner.clientList)
+      waitForClientToStart(client);
+    assertTrue(runner.awaitRunState(true, DELAY));
     runner.stop();
+    for (WaitingClient client : runner.clientList)
+      waitForClientToStop(client);
   }
 
+
   @Test
   public void testRunClientAndVisit() throws Exception {
-    executor.execute(runner);
-    MockClient client = new MockClient();
-    listener.addClient(client);
-    int retries = DELAY / SHORT_DELAY;
-    synchronized (client) {
-      while (retries-- > 0 && !client.isRunning()) {
-        client.wait(SHORT_DELAY);
-      }
-    }
+    listener.addSocket(new Socket());
+    // wait until all clients have been created
+    waitForRunnerClientListSizeToEqual(1);
+    WaitingClient client = runner.clientList.get(0);
+    waitForClientToStart(client);
+
     assertTrue(runner.awaitRunState(true, DELAY));
     MockClientVisitor visitor = new MockClientVisitor();
     runner.accept(visitor);
     assertSame(client, visitor.getLastVisited());
-    runner.stop();    
+    runner.stop();
+    waitForClientToStop(client);
+  }
+
+  private void waitForRunnerClientListSizeToEqual(int count) throws InterruptedException {
+    int retries = RETRIES;
+    while (retries-- > 0 && runner.clientList.size() != count) {
+      Thread.sleep(SHORT_DELAY);
+    }
   }
 
-  
-  static class InstrumentedConcurrentServerRunner 
-      extends ConcurrentServerRunner<MockClient> {
+  //  InstrumentedConcurrentServerRunner ================================================
+  static class InstrumentedConcurrentServerRunner extends ConcurrentServerRunner<WaitingClient> {
 
     private final Lock lock = new ReentrantLock();
     private final Condition runningCondition = lock.newCondition();
-    
-    public InstrumentedConcurrentServerRunner(
-        ServerListener<MockClient> listener, Executor executor) {
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/tony19/logback/file/2376051812/issues/source?bid=3222618&fileBranchId=3222711#l157)